Tools/TestWebKitAPI/Tests/WTF/WorkQueue.cpp

```diff
@@ -306,8 +306,78 @@ TEST(WTF_WorkQueue, DestroyDispatchedOnDispatchQueue)
         for (size_t j = 0; j < queueCount; ++j)
             queue[j]->dispatch([instance = std::make_unique<DestructionWorkQueueTester>(counter, *queue[j])] { }); // NOLINT
     }
+    // dispatchQOS() behaves the same as dispatch().
+    for (size_t i = 0; i < iterationCount; ++i) {
+        for (size_t j = 0; j < queueCount; ++j)
+            queue[j]->dispatchWithQOS([instance = std::make_unique<DestructionWorkQueueTester>(counter, *queue[j])] { }, Thread::QOS::UserInteractive); // NOLINT
+    }
     for (size_t j = 0; j < queueCount; ++j)
         queue[j]->dispatchSync([] { });
+    EXPECT_EQ(2u * queueCount * iterationCount, counter);
+}
+
+// Tests that the WorkQueue is kept alive for the duration of the work function.
+TEST(WTF_WorkQueue, DispatchWorkQueueKeptAliveDuringWorkFunction)
+{
+    std::atomic<size_t> counter = 0;
+    constexpr size_t queueCount = 50;
+    constexpr size_t iterationCount = 1000;
+    RefPtr<WorkQueue> queue[queueCount];
+    for (size_t i = 0; i < queueCount; ++i)
+        queue[i] = WorkQueue::create("com.apple.WebKit.Test.dispatchWorkQueueKeptAliveDuringWorkFunction", WorkQueue::QOS::UserInteractive);
+
+    for (size_t j = 0; j < queueCount; ++j) {
+        for (size_t i = 0; i < iterationCount; ++i) {
+            WorkQueue* myQueue = queue[j].get();
+            myQueue->dispatch([myQueue, &counter] {
+                // Use the reference. These would fail with ASAN unless the task holds the ref.
+#if USE(COCOA_EVENT_LOOP)
+                EXPECT_NOT_NULL(myQueue->dispatchQueue());
+#endif
+                sleep(0.01_s);
+                myQueue->dispatch([&counter] {
+                    ++counter;
+                });
+            });
+        }
+        queue[j] = nullptr;
+    }
+
```

Comment:
would it be worth checking the refcount of each queues ? in both the loops and at the end that they are all at 1 now (and that all tasks have run)

Reply:
The idea is that the refcounts are not 1 and that not all the tasks have run. That's the point of the test:
- to test the scenario where the ref is dropped
- the only ones holding the ref are the multiple pending tasks from within the framework
- the tasks also use the object without the owning ref, as they expect the framework to hold the ref